Systemd stage1 boot (initrd)

mbssrc opened the pull request:
<!--
    Copyright 2023 TII (SSRC) and the Ghaf contributors
    SPDX-License-Identifier: CC-BY-SA-4.0
-->

## Description of changes

This PR enables systemd in initrd for stage1 boot. For x1-carbon, enabled systemd `tpm`, `repart`, and `cryptsetup` for further work.

<!--
Summary of the proposed changes in the PR description in your own words. For dependency updates, please link to the changelog.
-->

## Checklist for things done

<!-- Please check, [X], to all that applies. Leave [ ] if an item does not apply but you have considered the check list item. Note that all of these are not hard requirements. They serve information to reviewers. When you fill the checklist, you indicate to reviewers you appreciate their work. -->

- [x] Summary of the proposed changes in the PR description
- [ ] More detailed description in the commit message(s)
- [x] Commits are squashed into relevant entities - avoid a lot of minimal dev time commits in the PR
- [ ] [Contribution guidelines](https://github.com/tiiuae/ghaf/blob/main/CONTRIBUTING.md) followed
- [ ] Ghaf documentation updated with the commit - https://tiiuae.github.io/ghaf/
- [ ] PR linked to architecture documentation and requirement(s) (ticket id)
- [ ] Test procedure described (or includes tests). Select one or more:
  - [x] Tested on Lenovo X1 `x86_64`
  - [x] Tested on Jetson Orin NX or AGX `aarch64`
  - [ ] Tested on Polarfire `riscv64`
- [ ] Author has run `nix flake check --accept-flake-config` and it passes
- [ ] All automatic Github Action checks pass - see [actions](https://github.com/tiiuae/ghaf/actions)
- [ ] Author has added reviewers and removed PR draft status

<!-- Additional description of omitted [ ] items if not obvious. -->

## Testing

Build and run on x1-carbon (gen11), generic-x86, vm, and Orin AGX.
Proper testing needed.

<!--
How this was tested by the author? How is this supposed to be tested
by people doing system testing?
-->

vunnyso commented on modules/microvm/virtualization/microvm/microvm-host.nix:
```diff
@@ -3,25 +3,30 @@
-        withNss = cfg.hostNetworkSupport;
-        withResolved = cfg.hostNetworkSupport;
+        withTpm2Tss = pkgs.stdenv.hostPlatform.isx86;
+        withRepart = true;
```
Can you please add `withFido2 = true;` which is needed while running `systemd-cryptenroll` otherwise will get `FIDO2 token support is not installed` error.

mbssrc committed:
Stage1 systemd boot (initrd)

Signed-off-by: Manuel Bluhm <manuel@ssrc.tii.ae>

mbssrc commented:
> Hi @mbssrc, will [postBootCmds](https://github.com/tiiuae/ghaf/blob/main/modules/disko/disko-basic-postboot.nix#L4-L59) will be executed as well once systemd is enabled in stage1?

Hey, yes it is being executed. My understanding is that it is set [here](https://github.com/NixOS/nixpkgs/blob/nixos-23.11/nixos/modules/system/boot/stage-2.nix#L24), then copied [here](https://github.com/NixOS/nixpkgs/blob/6f7a3ec8d35f32a26fcc5ea997a30561c10fd150/nixos/modules/system/activation/top-level.nix#L11), and executed [here](https://github.com/NixOS/nixpkgs/blob/6f7a3ec8d35f32a26fcc5ea997a30561c10fd150/nixos/modules/system/boot/systemd/initrd.nix#L561). You can run `journalctl -b` on the host to see it being executed as part of the `initrd-nixos-activation` service.

vilvo commented:
> > Hi @mbssrc, will [postBootCmds](https://github.com/tiiuae/ghaf/blob/main/modules/disko/disko-basic-postboot.nix#L4-L59) will be executed as well once systemd is enabled in stage1?
> 
> Hey, yes it is being executed. My understanding is that it is set [here](https://github.com/NixOS/nixpkgs/blob/nixos-23.11/nixos/modules/system/boot/stage-2.nix#L24), then copied [here](https://github.com/NixOS/nixpkgs/blob/6f7a3ec8d35f32a26fcc5ea997a30561c10fd150/nixos/modules/system/activation/top-level.nix#L11), and executed [here](https://github.com/NixOS/nixpkgs/blob/6f7a3ec8d35f32a26fcc5ea997a30561c10fd150/nixos/modules/system/boot/systemd/initrd.nix#L561). You can run `journalctl -b` on the host to see it being executed as part of the `initrd-nixos-activation` service.

Seems that setting [`IN_NIXOS_SYSTEMD_STAGE1` env variable to other than `true`](https://github.com/NixOS/nixpkgs/blob/bd22babc47eb6244acc46880aac65c8c82678051/nixos/modules/system/boot/stage-2-init.sh#L8) allows further debugging - if needed.

vunnyso commented:
> > > Hi @mbssrc, will [postBootCmds](https://github.com/tiiuae/ghaf/blob/main/modules/disko/disko-basic-postboot.nix#L4-L59) will be executed as well once systemd is enabled in stage1?
> > 
> > 
> > Hey, yes it is being executed. My understanding is that it is set [here](https://github.com/NixOS/nixpkgs/blob/nixos-23.11/nixos/modules/system/boot/stage-2.nix#L24), then copied [here](https://github.com/NixOS/nixpkgs/blob/6f7a3ec8d35f32a26fcc5ea997a30561c10fd150/nixos/modules/system/activation/top-level.nix#L11), and executed [here](https://github.com/NixOS/nixpkgs/blob/6f7a3ec8d35f32a26fcc5ea997a30561c10fd150/nixos/modules/system/boot/systemd/initrd.nix#L561). You can run `journalctl -b` on the host to see it being executed as part of the `initrd-nixos-activation` service.
> 
> Seems that setting [`IN_NIXOS_SYSTEMD_STAGE1` env variable to other than `true`](https://github.com/NixOS/nixpkgs/blob/bd22babc47eb6244acc46880aac65c8c82678051/nixos/modules/system/boot/stage-2-init.sh#L8) allows further debugging - if needed.

Thanks @vilvo and @mbssrc for your response.

vilvo commented:
Seems `init` is missing.
![1](https://github.com/tiiuae/ghaf/assets/1027150/ca288deb-216f-48bd-bb37-c23b0fbedb6c)
![2](https://github.com/tiiuae/ghaf/assets/1027150/e31a503a-3aea-4ea2-ba6a-f645965e00d6)

mbssrc commented:
> Seems `init` is missing. ![1](https://private-user-images.githubusercontent.com/1027150/319890519-ca288deb-216f-48bd-bb37-c23b0fbedb6c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIzMDExMDIsIm5iZiI6MTcxMjMwMDgwMiwicGF0aCI6Ii8xMDI3MTUwLzMxOTg5MDUxOS1jYTI4OGRlYi0yMTZmLTQ4YmQtYmIzNy1jMjNiMGZiZWRiNmMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MDVUMDcwNjQyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Mzc3OGJjN2Q2NDRlZmJlMDFhM2M4NDhhNDBlZWZhNjQzMjkwYjVmZjNmZWZlZDhlNTljY2U5ODA3OWJlMmQyMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.E1PI-94YU6zUo2p7mG1JB-78FeiIB7WyyTjE7P54oFM) ![2](https://private-user-images.githubusercontent.com/1027150/319890534-e31a503a-3aea-4ea2-ba6a-f645965e00d6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIzMDExMDIsIm5iZiI6MTcxMjMwMDgwMiwicGF0aCI6Ii8xMDI3MTUwLzMxOTg5MDUzNC1lMzFhNTAzYS0zYWVhLTRlYTItYmE2YS1mNjQ1OTY1ZTAwZDYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MDVUMDcwNjQyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZGYxMzFmZDgyOGIxYTUwZjI0NzU3YTc3OTQxYmE5NmYwMGRjOWEyNWIzZjNkZWY1MWJlZTk2NjQ1YmU1NTkzZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.1WLRlkvm0CX6KucBQdwYcve53uNyuM4Dy0OUr4_ZEU8)

Is this the same problem that we had before, where you have a version installed on the laptop and try to boot from SD?

vilvo commented:
> Is this the same problem that we had before, where you have a version installed on the laptop and try to boot from SD?

Was. `dd if=/dev/zero of=/dev/nvme0n1` in the maintenance shell to the rescue. Sorry about that. Boots to the desktop now.

Commit:
fix: DMAR: Unknown option - 'igx_off,sm_on'

* fixing the typo s/igx_off/igfx_off/ to
"DMAR: Disable GFX device mapping" allows host to boot
 but it does not boot to the desktop with graphics
* fix is to remove the invalid parameter
* tested on Lenovo X1 gen11

Signed-off-by: Ville Ilvonen <ville.ilvonen@unikie.com>

vilvo commented:
@leivos-unikie - can you please organize proper regression testing for this?

leivos-unikie commented:
> @leivos-unikie - can you please organize proper regression testing for this?

@barnabakos messaged yesterday that he would test this. Barna have you started this?

leivos-unikie commented:
build3 remote builder has been down recently

Both @barnabakos and me have tried to build for lenovo-x1 locally but we get hash mismatch
       > copying path '/nix/store/pj0xc0bx70aly4gvcbcawbkf71zg2009-unit-microvm-chromium-vm.service' to 'local'...
       > error: hash mismatch importing path '/nix/store/q53arplxmgs085gqa62s8j2kj4na8ksd-qemu-8.1.3';
 
 Barna tested on Orin AGX and there was no new issues found.

milva-unikie commented:
Build3 remote builder is back up and building successful Lenovo-X1 images.

Everything is working as expected with X1.

leivos-unikie commented:
Tested vm-debug target
- Apps launch (except Element)
- Applicable ci-test-automation tests pass

brianmcgillion commented:
> Tested vm-debug target
> 
> * Apps launch (except Element)
Should element work? I.e. does it work in current image?

leivos-unikie commented:
> > Tested vm-debug target
> > 
> > * Apps launch (except Element)
> >   Should element work? I.e. does it work in current image?

Element didn't work in vm-debug even before this PR.
